File: tests/oracle/userop_localtype_metadata_1204.cc
/*
 * USEROP-LOCALTYPE-METADATA-0001: locked Ghidra 12.0.4 oracle for
 * DatatypeUserOp metadata and UserOpManage registration/query behavior.
 *
 * The projection observes factory pointer identity, slot-minus-one lookup,
 * null-input compaction, descriptor replacement, duplicate builtin identity,
 * and the exact no-mutation state after each registration exception.
 */
#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "userop.hh"

#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace std;
using namespace ghidra;

// C++11 private-member fixture bridge.  UserOpManage::registerOp is private,
// but it is the production function whose ordering must be observed.  Explicit
// template instantiation is the standard compile-time access bridge; the
// fixture invokes the unmodified locked function body.
template<typename Tag,typename Tag::type member>
struct PrivateMemberAccess {
  friend typename Tag::type access(Tag) { return member; }
};

struct RegisterOpTag {
  typedef void (UserOpManage::*type)(UserPcodeOp *);
  friend type access(RegisterOpTag);
};

template struct PrivateMemberAccess<RegisterOpTag,&UserOpManage::registerOp>;

void registerOp(UserOpManage &manager,UserPcodeOp *op)

{
  (manager.*access(RegisterOpTag()))(op);
}

string registerResult(UserOpManage &manager,UserPcodeOp *op)

{
  try {
    registerOp(manager,op);
    return "NO_ERROR";
  }
  catch(const LowlevelError &err) {
    delete op;                 // registerOp did not take ownership on these paths
    return err.explain;
  }
}

string badBuiltinResult(UserOpManage &manager,uint4 id)

{
  try {
    manager.registerBuiltin(id);
    return "NO_ERROR";
  }
  catch(const LowlevelError &err) {
    return err.explain;
  }
}

void emitBool(const string &key,bool value)

{
  cout << key << '=' << (value ? 1 : 0) << '\n';
}

void emitDatatypeBuiltin(Architecture &arch,uint4 id,const string &key,
                         Datatype *expectedOut,Datatype *expectedIn0,
                         Datatype *expectedIn1,Datatype *expectedIn2)

{
  UserPcodeOp *descriptor = arch.userops.registerBuiltin(id);
  cout << key << ".type=" << descriptor->getType() << '\n';
  cout << key << ".index=" << descriptor->getIndex() << '\n';
  emitBool(key + ".manager_same",arch.userops.getOp(id) == descriptor);
  emitBool(key + ".out",descriptor->getOutputLocal((PcodeOp *)0) == expectedOut);
  emitBool(key + ".slot0_null",descriptor->getInputLocal((PcodeOp *)0,0) == (Datatype *)0);
  emitBool(key + ".slot1",descriptor->getInputLocal((PcodeOp *)0,1) == expectedIn0);
  emitBool(key + ".slot2",descriptor->getInputLocal((PcodeOp *)0,2) == expectedIn1);
  emitBool(key + ".slot3",descriptor->getInputLocal((PcodeOp *)0,3) == expectedIn2);
  emitBool(key + ".slot4_null",descriptor->getInputLocal((PcodeOp *)0,4) == (Datatype *)0);
}

void runFixture(const string &specDirectory,const string &binary)

{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  BfdArchitecture arch(binary,"default",&cerr);
  DocumentStorage store;
  arch.init(store);
  if (arch.archid != "x86:LE:64:default:gcc")
    throw runtime_error("runtime architecture/compiler drifted: " + arch.archid);

  TypeFactory *factory = arch.types;
  const int4 pointerSize = factory->getSizeOfPointer();
  const int4 wordSize = arch.getDefaultDataSpace()->getWordSize();
  Datatype *voidType = factory->getTypeVoid();
  Datatype *voidPointer = factory->getTypePointer(pointerSize,voidType,wordSize);
  Datatype *intType = factory->getBase(4,TYPE_INT);
  Datatype *charType = factory->getTypeChar(factory->getSizeOfChar());
  Datatype *charPointer = factory->getTypePointer(pointerSize,charType,wordSize);
  Datatype *wcharType = factory->getTypeChar(factory->getSizeOfWChar());
  Datatype *wcharPointer = factory->getTypePointer(pointerSize,wcharType,wordSize);

  emitDatatypeBuiltin(arch,UserPcodeOp::BUILTIN_MEMCPY,"builtin.memcpy",
      voidPointer,voidPointer,voidPointer,intType);
  UserPcodeOp *memcpyDescriptor = arch.userops.getOp(UserPcodeOp::BUILTIN_MEMCPY);
  UserPcodeOp *memcpyRepeat = arch.userops.registerBuiltin(UserPcodeOp::BUILTIN_MEMCPY);
  emitBool("builtin.memcpy.repeat_same",memcpyRepeat == memcpyDescriptor);
  emitBool("builtin.memcpy.repeat_out",
      memcpyRepeat->getOutputLocal((PcodeOp *)0) == voidPointer);

  emitDatatypeBuiltin(arch,UserPcodeOp::BUILTIN_STRNCPY,"builtin.strncpy",
      charPointer,charPointer,charPointer,intType);
  emitDatatypeBuiltin(arch,UserPcodeOp::BUILTIN_WCSNCPY,"builtin.wcsncpy",
      wcharPointer,wcharPointer,wcharPointer,intType);
  emitBool("builtin.memcpy.stable_after_growth",
      arch.userops.getOp(UserPcodeOp::BUILTIN_MEMCPY) == memcpyDescriptor);

  const uint4 badId = UserPcodeOp::BUILTIN_WCSNCPY + 1;
  cout << "builtin.bad.error=" << badBuiltinResult(arch.userops,badId) << '\n';
  emitBool("builtin.bad.absent",arch.userops.getOp(badId) == (UserPcodeOp *)0);
  emitBool("builtin.memcpy.stable_after_error",
      arch.userops.getOp(UserPcodeOp::BUILTIN_MEMCPY) == memcpyDescriptor);

  UserOpManage manager;
  registerOp(manager,new UnspecializedPcodeOp("typed",&arch,2));
  emitBool("custom.gap0_null",manager.getOp(0) == (UserPcodeOp *)0);
  emitBool("custom.gap1_null",manager.getOp(1) == (UserPcodeOp *)0);
  emitBool("custom.base_out_null",
      manager.getOp(2)->getOutputLocal((PcodeOp *)0) == (Datatype *)0);
  emitBool("custom.base_in_null",
      manager.getOp(2)->getInputLocal((PcodeOp *)0,1) == (Datatype *)0);

  cout << "custom.typed.error=" << registerResult(manager,
      new DatatypeUserOp("typed",&arch,2,voidPointer,
          (Datatype *)0,charType,(Datatype *)0,intType)) << '\n';
  UserPcodeOp *typed = manager.getOp(2);
  cout << "custom.typed.type=" << typed->getType() << '\n';
  emitBool("custom.by_name_same",manager.getOp("typed") == typed);
  emitBool("custom.out",typed->getOutputLocal((PcodeOp *)0) == voidPointer);
  emitBool("custom.slot0_null",typed->getInputLocal((PcodeOp *)0,0) == (Datatype *)0);
  emitBool("custom.slot1_compacted",typed->getInputLocal((PcodeOp *)0,1) == charType);
  emitBool("custom.slot2_compacted",typed->getInputLocal((PcodeOp *)0,2) == intType);
  emitBool("custom.slot3_null",typed->getInputLocal((PcodeOp *)0,3) == (Datatype *)0);

  cout << "custom.replace.error=" << registerResult(manager,
      new DatatypeUserOp("typed",&arch,2,intType,voidType)) << '\n';
  typed = manager.getOp(2);
  emitBool("custom.replace.out",typed->getOutputLocal((PcodeOp *)0) == intType);
  emitBool("custom.replace.slot1",typed->getInputLocal((PcodeOp *)0,1) == voidType);
  emitBool("custom.replace.slot2_null",
      typed->getInputLocal((PcodeOp *)0,2) == (Datatype *)0);

  cout << "custom.same_name_new_index.error=" << registerResult(manager,
      new DatatypeUserOp("typed",&arch,3,voidType,intType)) << '\n';
  emitBool("custom.same_name_new_index.gap_preserved",
      manager.getOp(3) == (UserPcodeOp *)0);
  emitBool("custom.same_name_new_index.old_preserved",
      manager.getOp(2)->getOutputLocal((PcodeOp *)0) == intType);

  cout << "custom.same_index_new_name.error=" << registerResult(manager,
      new DatatypeUserOp("other",&arch,2,voidType,intType)) << '\n';
  emitBool("custom.same_index_new_name.absent",
      manager.getOp("other") == (UserPcodeOp *)0);
  emitBool("custom.same_index_new_name.old_preserved",
      manager.getOp("typed") == manager.getOp(2));

  cout << "custom.negative.error=" << registerResult(manager,
      new DatatypeUserOp("negative",&arch,-1,voidType,intType)) << '\n';
  emitBool("custom.negative.absent",
      manager.getOp("negative") == (UserPcodeOp *)0);

  cout << "custom.missing.error=" << registerResult(manager,
      new DatatypeUserOp("missing",&arch,4,(Datatype *)0)) << '\n';
  emitBool("custom.missing.out_null",
      manager.getOp(4)->getOutputLocal((PcodeOp *)0) == (Datatype *)0);
  emitBool("custom.missing.in_null",
      manager.getOp(4)->getInputLocal((PcodeOp *)0,1) == (Datatype *)0);
  emitBool("custom.error_gap_still_null",manager.getOp(3) == (UserPcodeOp *)0);
}

} // namespace

int main(int argc,char **argv)

{
  try {
    if (argc != 3)
      throw invalid_argument(
          "usage: userop_localtype_metadata_1204 SPEC_DIRECTORY BINARY");
    runFixture(argv[1],argv[2]);
    return 0;
  }
  catch(const LowlevelError &err) {
    cerr << "LowlevelError: " << err.explain << '\n';
  }
  catch(const exception &err) {
    cerr << "std::exception: " << err.what() << '\n';
  }
  return 1;
}

File: tests/oracle/userop_localtype_metadata_1204.rs
// USEROP-LOCALTYPE-METADATA-0001: Rust comparand for the locked Ghidra
// DatatypeUserOp/UserOpManage fixture.  Every Arc comes directly from one
// TypeFactory; ptr_eq is the Rust observation of Ghidra Datatype* identity.

use std::sync::Arc;

use rugra::type_system::datatype::{Datatype, TypeMetatype};
use rugra::type_system::typefactory::{SizeArchInputs, TypeFactory};
use rugra::userop::{
    DatatypeUserOp, UserOpManage, UserOpType, UserPcodeOp, BUILTIN_MEMCPY, BUILTIN_STRNCPY,
    BUILTIN_WCSNCPY,
};

fn emit_bool(key: &str, value: bool) {
    println!("{key}={}", if value { 1 } else { 0 });
}

fn local_eq(actual: Option<&Arc<Datatype>>, expected: &Arc<Datatype>) -> bool {
    actual.is_some_and(|datatype| Arc::ptr_eq(datatype, expected))
}

fn emit_datatype_builtin(
    manager: &UserOpManage,
    id: u32,
    key: &str,
    expected_out: &Arc<Datatype>,
    expected_in0: &Arc<Datatype>,
    expected_in1: &Arc<Datatype>,
    expected_in2: &Arc<Datatype>,
) {
    let descriptor = manager.get_op(id as i32).expect("typed builtin descriptor");
    println!("{key}.type={}", descriptor.get_type() as i32);
    println!("{key}.index={}", descriptor.get_index());
    emit_bool(
        &format!("{key}.manager_same"),
        std::ptr::eq(
            descriptor,
            manager
                .get_op(id as i32)
                .expect("same typed builtin descriptor"),
        ),
    );
    emit_bool(
        &format!("{key}.out"),
        local_eq(descriptor.get_output_local(), expected_out),
    );
    emit_bool(
        &format!("{key}.slot0_null"),
        descriptor.get_input_local(0).is_none(),
    );
    emit_bool(
        &format!("{key}.slot1"),
        local_eq(descriptor.get_input_local(1), expected_in0),
    );
    emit_bool(
        &format!("{key}.slot2"),
        local_eq(descriptor.get_input_local(2), expected_in1),
    );
    emit_bool(
        &format!("{key}.slot3"),
        local_eq(descriptor.get_input_local(3), expected_in2),
    );
    emit_bool(
        &format!("{key}.slot4_null"),
        descriptor.get_input_local(4).is_none(),
    );
}

fn main() {
    let mut factory = TypeFactory::new(8);
    factory.setup_sizes(&SizeArchInputs {
        stack_spacebase_size: Some(8),
        default_data_space_addr_size: 8,
        default_size: 8,
        far_pointer: None,
    });

    let void_type = factory.get_type_void();
    let void_pointer = factory.get_type_pointer(8, void_type.clone(), 1);
    let int_type = factory
        .get_base(4, TypeMetatype::Int)
        .expect("canonical int type");
    let char_type = factory.get_type_char(factory.get_size_of_char() as usize);
    let char_pointer = factory.get_type_pointer(8, char_type.clone(), 1);
    let wchar_type = factory.get_type_char(factory.get_size_of_wchar() as usize);
    let wchar_pointer = factory.get_type_pointer(8, wchar_type.clone(), 1);

    let mut builtins = UserOpManage::new();
    builtins
        .register_builtin_with_local_types(
            BUILTIN_MEMCPY,
            Some(void_pointer.clone()),
            vec![
                Some(void_pointer.clone()),
                Some(void_pointer.clone()),
                Some(int_type.clone()),
            ],
        )
        .expect("register memcpy");
    emit_datatype_builtin(
        &builtins,
        BUILTIN_MEMCPY,
        "builtin.memcpy",
        &void_pointer,
        &void_pointer,
        &void_pointer,
        &int_type,
    );
    let memcpy_descriptor = builtins
        .get_op(BUILTIN_MEMCPY as i32)
        .expect("memcpy descriptor") as *const UserPcodeOp;
    builtins
        .register_builtin_with_local_types(
            BUILTIN_MEMCPY,
            Some(int_type.clone()),
            vec![Some(int_type.clone())],
        )
        .expect("repeat memcpy");
    emit_bool(
        "builtin.memcpy.repeat_same",
        std::ptr::eq(
            memcpy_descriptor,
            builtins
                .get_op(BUILTIN_MEMCPY as i32)
                .expect("repeated memcpy descriptor") as *const UserPcodeOp,
        ),
    );
    emit_bool(
        "builtin.memcpy.repeat_out",
        local_eq(
            builtins.get_output_local(BUILTIN_MEMCPY as i32),
            &void_pointer,
        ),
    );

    builtins
        .register_builtin_with_local_types(
            BUILTIN_STRNCPY,
            Some(char_pointer.clone()),
            vec![
                Some(char_pointer.clone()),
                Some(char_pointer.clone()),
                Some(int_type.clone()),
            ],
        )
        .expect("register strncpy");
    emit_datatype_builtin(
        &builtins,
        BUILTIN_STRNCPY,
        "builtin.strncpy",
        &char_pointer,
        &char_pointer,
        &char_pointer,
        &int_type,
    );
    builtins
        .register_builtin_with_local_types(
            BUILTIN_WCSNCPY,
            Some(wchar_pointer.clone()),
            vec![
                Some(wchar_pointer.clone()),
                Some(wchar_pointer.clone()),
                Some(int_type.clone()),
            ],
        )
        .expect("register wcsncpy");
    emit_datatype_builtin(
        &builtins,
        BUILTIN_WCSNCPY,
        "builtin.wcsncpy",
        &wchar_pointer,
        &wchar_pointer,
        &wchar_pointer,
        &int_type,
    );
    emit_bool(
        "builtin.memcpy.stable_after_growth",
        std::ptr::eq(
            memcpy_descriptor,
            builtins
                .get_op(BUILTIN_MEMCPY as i32)
                .expect("stable memcpy descriptor") as *const UserPcodeOp,
        ),
    );

    let bad_id = BUILTIN_WCSNCPY + 1;
    let bad_error = builtins
        .try_register_builtin_by_id(bad_id)
        .expect_err("bad builtin id must fail");
    println!("builtin.bad.error={bad_error}");
    emit_bool(
        "builtin.bad.absent",
        builtins.get_op(bad_id as i32).is_none(),
    );
    emit_bool(
        "builtin.memcpy.stable_after_error",
        std::ptr::eq(
            memcpy_descriptor,
            builtins
                .get_op(BUILTIN_MEMCPY as i32)
                .expect("memcpy after error") as *const UserPcodeOp,
        ),
    );

    let mut manager = UserOpManage::new();
    manager
        .register_user_op(UserPcodeOp::new(
            "typed".to_string(),
            UserOpType::Unspecialized,
            2,
        ))
        .expect("register unspecialized typed op");
    emit_bool("custom.gap0_null", manager.get_op(0).is_none());
    emit_bool("custom.gap1_null", manager.get_op(1).is_none());
    emit_bool(
        "custom.base_out_null",
        manager.get_output_local(2).is_none(),
    );
    emit_bool(
        "custom.base_in_null",
        manager.get_input_local(2, 1).is_none(),
    );

    let typed_result = manager.register_datatype_user_op(DatatypeUserOp::new(
        "typed".to_string(),
        2,
        Some(void_pointer.clone()),
        vec![None, Some(char_type.clone()), None, Some(int_type.clone())],
    ));
    println!(
        "custom.typed.error={}",
        typed_result.err().unwrap_or_else(|| "NO_ERROR".to_string())
    );
    let typed = manager.get_op(2).expect("typed descriptor");
    println!("custom.typed.type={}", typed.get_type() as i32);
    emit_bool(
        "custom.by_name_same",
        std::ptr::eq(
            manager.get_op_by_name("typed").expect("typed name lookup"),
            typed,
        ),
    );
    emit_bool(
        "custom.out",
        local_eq(manager.get_output_local(2), &void_pointer),
    );
    emit_bool("custom.slot0_null", manager.get_input_local(2, 0).is_none());
    emit_bool(
        "custom.slot1_compacted",
        local_eq(manager.get_input_local(2, 1), &char_type),
    );
    emit_bool(
        "custom.slot2_compacted",
        local_eq(manager.get_input_local(2, 2), &int_type),
    );
    emit_bool("custom.slot3_null", manager.get_input_local(2, 3).is_none());

    let replace_result = manager.register_datatype_user_op(DatatypeUserOp::new(
        "typed".to_string(),
        2,
        Some(int_type.clone()),
        vec![Some(void_type.clone())],
    ));
    println!(
        "custom.replace.error={}",
        replace_result
            .err()
            .unwrap_or_else(|| "NO_ERROR".to_string())
    );
    emit_bool(
        "custom.replace.out",
        local_eq(manager.get_output_local(2), &int_type),
    );
    emit_bool(
        "custom.replace.slot1",
        local_eq(manager.get_input_local(2, 1), &void_type),
    );
    emit_bool(
        "custom.replace.slot2_null",
        manager.get_input_local(2, 2).is_none(),
    );

    let same_name_error = manager
        .register_datatype_user_op(DatatypeUserOp::new(
            "typed".to_string(),
            3,
            Some(void_type.clone()),
            vec![Some(int_type.clone())],
        ))
        .expect_err("same name at new index must fail");
    println!("custom.same_name_new_index.error={same_name_error}");
    emit_bool(
        "custom.same_name_new_index.gap_preserved",
        manager.get_op(3).is_none(),
    );
    emit_bool(
        "custom.same_name_new_index.old_preserved",
        local_eq(manager.get_output_local(2), &int_type),
    );

    let same_index_error = manager
        .register_datatype_user_op(DatatypeUserOp::new(
            "other".to_string(),
            2,
            Some(void_type.clone()),
            vec![Some(int_type.clone())],
        ))
        .expect_err("new name at same index must fail");
    println!("custom.same_index_new_name.error={same_index_error}");
    emit_bool(
        "custom.same_index_new_name.absent",
        manager.get_op_by_name("other").is_none(),
    );
    emit_bool(
        "custom.same_index_new_name.old_preserved",
        std::ptr::eq(
            manager
                .get_op_by_name("typed")
                .expect("preserved typed name"),
            manager.get_op(2).expect("preserved typed index"),
        ),
    );

    let negative_error = manager
        .register_datatype_user_op(DatatypeUserOp::new(
            "negative".to_string(),
            -1,
            Some(void_type.clone()),
            vec![Some(int_type.clone())],
        ))
        .expect_err("negative index must fail");
    println!("custom.negative.error={negative_error}");
    emit_bool(
        "custom.negative.absent",
        manager.get_op_by_name("negative").is_none(),
    );

    let missing_result = manager.register_datatype_user_op(DatatypeUserOp::new(
        "missing".to_string(),
        4,
        None,
        Vec::new(),
    ));
    println!(
        "custom.missing.error={}",
        missing_result
            .err()
            .unwrap_or_else(|| "NO_ERROR".to_string())
    );
    emit_bool(
        "custom.missing.out_null",
        manager.get_output_local(4).is_none(),
    );
    emit_bool(
        "custom.missing.in_null",
        manager.get_input_local(4, 1).is_none(),
    );
    emit_bool("custom.error_gap_still_null", manager.get_op(3).is_none());
}
